x/incentive/keeper/rewards.go

```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
 	claim.Reward = claim.Reward.Add(newRewardsCoin)
-	k.SetClaim(ctx, claim)
+	k.SetUSDXMintingClaim(ctx, claim)
 	return
 }
 
+// InitializeHardSupplyReward initializes the supply-side of a hard liquidity provider claim
+// by creating the claim and setting the supply reward factor index
+func (k Keeper) InitializeHardSupplyReward(ctx sdk.Context, deposit hardtypes.Deposit) {
+	_, claimFound := k.GetHardLiquidityProviderClaim(ctx, deposit.Depositor)
+	if claimFound {
+		return
+	}
+
+	var supplyRewardIndexes types.RewardIndexes
+	for _, coin := range deposit.Amount {
+		_, found := k.GetHardSupplyRewardPeriod(ctx, coin.Denom)
+		if !found {
+			continue
+		}
+
+		supplyFactor, foundSupplyFactor := k.GetHardSupplyRewardFactor(ctx, coin.Denom)
+		if !foundSupplyFactor {
+			supplyFactor = sdk.ZeroDec()
+		}
+
+		supplyRewardIndexes = append(supplyRewardIndexes, types.NewRewardIndex(coin.Denom, supplyFactor))
+	}
+
+	claim, found := k.GetHardLiquidityProviderClaim(ctx, deposit.Depositor)
+	if !found {
+		claim = types.NewHardLiquidityProviderClaim(deposit.Depositor,
+			sdk.NewCoin(types.HardLiquidityRewardDenom, sdk.ZeroInt()),
+			supplyRewardIndexes, nil, nil)
+		k.SetHardLiquidityProviderClaim(ctx, claim)
+		return
+	}
+
+	// User has a HARD claim, but supply reward indexes are empty
+	claim.SupplyRewardIndexes = supplyRewardIndexes
+	k.SetHardLiquidityProviderClaim(ctx, claim)
+}
+
+// SynchronizeHardSupplyReward updates the claim object by adding any accumulated rewards
+// and updating the reward index value
+func (k Keeper) SynchronizeHardSupplyReward(ctx sdk.Context, deposit hardtypes.Deposit) {
+	claim, found := k.GetHardLiquidityProviderClaim(ctx, deposit.Depositor)
+	if !found {
+		return
+	}
+
+	for _, coin := range deposit.Amount {
+		supplyFactor, found := k.GetHardSupplyRewardFactor(ctx, coin.Denom)
+		if !found {
+			continue
+		}
+
+		supplyIndex, hasSupplyRewardIndex := claim.HasSupplyRewardIndex(coin.Denom)
+		if !hasSupplyRewardIndex {
+			continue
+		}
+
+		userRewardFactor := claim.SupplyRewardIndexes[supplyIndex].RewardFactor
+		rewardsAccumulatedFactor := supplyFactor.Sub(userRewardFactor)
+		if rewardsAccumulatedFactor.IsZero() {
+			continue
+		}
+		claim.SupplyRewardIndexes[supplyIndex].RewardFactor = supplyFactor
+
+		newRewardsAmount := rewardsAccumulatedFactor.Mul(deposit.Amount.AmountOf(coin.Denom).ToDec()).RoundInt()
+		if newRewardsAmount.IsZero() {
+			continue
+		}
+
+		newRewardsCoin := sdk.NewCoin(types.HardLiquidityRewardDenom, newRewardsAmount)
+		claim.Reward = claim.Reward.Add(newRewardsCoin)
+	}
+
+	k.SetHardLiquidityProviderClaim(ctx, claim)
+}
+
+// InitializeHardBorrowReward initializes the borrow-side of a hard liquidity provider claim
+// by creating the claim and setting the borrow reward factor index
+func (k Keeper) InitializeHardBorrowReward(ctx sdk.Context, borrow hardtypes.Borrow) {
+	claim, claimFound := k.GetHardLiquidityProviderClaim(ctx, borrow.Borrower)
+	if !claimFound {
+		return
```

Comment:
Imagine this scenario:

1. User deposits coinA, which we don't have incentives for - a claim will not be created.
2. User borrows coinB, which we do have incentives for - claim doesn't exist!

Again, if I have the state-machine correct, it's not safe to return here. If a claim doesn't exist:

```
if !claimFound {
  claim = types.NewHardLiquidityProviderClaim(deposit.Depositor,
			sdk.NewCoin(types.HardLiquidityRewardDenom, sdk.ZeroInt()),
			nil, nil, nil)
}
```

Reply:
Updated in https://github.com/Kava-Labs/kava/pull/757/commits/e300ada916f042f11e21e6d0c37e49fedb388380